wasm/derive/src/lib.rs

```diff
@@ -1,89 +1,229 @@
-//! Macros for writing smartcontracts
+//! Macros for writing smart contracts
 
 #![allow(clippy::str_to_string)]
 
 use proc_macro::TokenStream;
-use proc_macro_error::{abort, proc_macro_error};
 use quote::quote;
-use syn::{parse_macro_input, parse_quote, ItemFn, Path, ReturnType, Signature, Type};
+use syn::{parse_macro_input, parse_quote, punctuated::Punctuated};
 
-/// Used to annotate user-defined function which starts the execution of smartcontract
-#[proc_macro_error]
+mod kw {
+    syn::custom_keyword!(params);
+
+    pub mod param_types {
+        syn::custom_keyword!(authority);
+        syn::custom_keyword!(triggering_event);
+    }
+}
+
+/// Enum representing possible attributes for [`entrypoint`] macro
+enum Attr {
+    Params(ParamsAttr),
+    Empty,
+}
+
+impl syn::parse::Parse for Attr {
+    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
+        if input.is_empty() {
+            return Ok(Attr::Empty);
+        }
+
+        Ok(Attr::Params(input.parse()?))
+    }
+}
+
+/// Attribute with expected parameters for smart contract entrypoint function
+struct ParamsAttr {
+    _params_kw: kw::params,
```

Comment:
why the underscores in front?

Reply:
Because this field is not used anywhere and clippy complains about it. As far as I remember it's a common practice for structures like that (i.e. `actor/derive/` and `config/derive/`)